x-pack/solutions/security/plugins/security_solution/public/detection_engine/rule_creation/components/threat_match_mapping_edit/threat_match_mapping_field.tsx

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useCallback, useEffect } from 'react';
+import { EuiFormRow } from '@elastic/eui';
+import type { DataViewBase } from '@kbn/es-query';
+import { createOrNewEntryItem } from '../../../../common/components/threat_match/helpers';
+import type { ThreatMapEntries } from '../../../../common/components/threat_match/types';
+import { ThreatMatchComponent } from '../../../../common/components/threat_match';
+import type { FieldHook } from '../../../../shared_imports';
+import { getFieldValidityAndErrorMessage } from '../../../../shared_imports';
+
+export const DEFAULT_THREAT_MAPPING_VALUE = [createOrNewEntryItem()];
+
+interface ThreatMatchMappingFieldProps {
+  field: FieldHook<ThreatMapEntries[]>;
+  indexPatterns: DataViewBase;
+  threatIndexPatterns: DataViewBase;
+}
+
+export function ThreatMatchMappingField({
+  field,
+  indexPatterns,
+  threatIndexPatterns,
+}: ThreatMatchMappingFieldProps): JSX.Element {
+  const { isInvalid, errorMessage } = getFieldValidityAndErrorMessage(field);
+  const { setValue, validate } = field;
+
+  // We have to make sure validation runs against the latest source events index patterns
+  // and threat match index patterns.
+  // Form lib's `fieldsToValidateOnChange` on the corresponding index patterns edit fields
+  // doesn't help here. It leads to running threat match mapping validation before render
+  // of this component happens. In the end validation runs against previous index patterns
+  // producing invalid validation results.
+  //
+  // Validating the field imperatively here fixes this issue.
+  //
+  // Additional pitfall here is `validate` function changing its reference every render. Including it
+  // in useEffect's deps leads to infinite re-render.
+  useEffect(() => {
+    validate();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [indexPatterns, threatIndexPatterns]);
```

Comment:
This part looks sub-optimal but it's the best I was able to make to re-run threat match mapping validation upon index pattern changes. An alternative is using `formData` in validation to read index patterns/data source, transform them to data view asynchronously and then validate fields against known fields. Alternative's drawback is difficulty to reuse it in rule upgrade workflow.

Any bright ideas are welcome.